admin/test/test_release.py

```diff
@@ -661,3 +667,637 @@ def test_publish_non_release_fails(self):
+        """
+        existing_s3_keys = {
+            os.path.join(self.target_key, 'existing_package.rpm'): '',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys,
+            },
+        )
+
+        unspecified_package = 'unspecified-package-0.3.3-0.dev.7.noarch.rpm'
+        repo_contents = {
+            'clusterhq-flocker-cli-0.3.3-0.dev.7.noarch.rpm': '',
+            'clusterhq-flocker-node-0.3.3-0.dev.7.noarch.rpm': '',
+            unspecified_package: '',
+        }
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=repo_contents),
+            packages=self.packages,
+        )
+
+        self.assertNotIn(
+            os.path.join(self.target_key, unspecified_package),
+            aws.s3_buckets[self.target_bucket])
+
+    def test_development_repositories_created(self):
+        """
+        Calling :func:`upload_rpms` creates development repositories for
+        CentOS 7 and Fedora 20 for a development release.
+        """
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: {},
+            },
+        )
+
+        repo_contents = {
+            'results/omnibus/0.3.3dev7/fedora-20/clusterhq-flocker-cli-0.3.3-0.dev.7.noarch.rpm': '',  # noqa
+            'results/omnibus/0.3.3dev7/fedora-20/clusterhq-flocker-node-0.3.3-0.dev.7.noarch.rpm': '',  # noqa
+            'results/omnibus/0.3.3dev7/centos-7/clusterhq-flocker-cli-0.3.3-0.dev.7.noarch.rpm': '',  # noqa
+            'results/omnibus/0.3.3dev7/centos-7/clusterhq-flocker-node-0.3.3-0.dev.7.noarch.rpm': '',  # noqa
+
+        }
+
+        self.upload_rpms(
+            aws=aws,
+            yum=FakeYum(),
+            scratch_directory=self.scratch_directory,
+            target_bucket=self.target_bucket,
+            version='0.3.3dev7',
+            build_server=self.create_fake_repository(files=repo_contents),
+        )
+
+        expected_files = set()
+        for operating_system in self.operating_systems:
+            for file in [
+                'clusterhq-flocker-cli-0.3.3-0.dev.7.noarch.rpm',
+                'clusterhq-flocker-node-0.3.3-0.dev.7.noarch.rpm',
+                'repodata/repomd.xml',
+            ]:
+                path = os.path.join(
+                    'development',
+                    operating_system['distro'],
+                    operating_system['version'],
+                    operating_system['arch'],
+                    file,
+                )
+                expected_files.add(path)
+
+        files_on_s3 = aws.s3_buckets[self.target_bucket].keys()
+        self.assertTrue(expected_files.issubset(set(files_on_s3)))
```

Comment:
Tests with "issubset" fail just telling you that False does not equal True. Mitigation options for this problem include:

* Put a descriptive message in
* Compare sorted lists instead of sets

Any preference / suggestion @tomprince ?